### PR TITLE
topology: add dmic channel maps for different channel counts

### PR DESCRIPTION
**Pull request by juimonen**
Volume and switch controls in the dmic pipeline need different channel
maps based on dmic channel count. This will enable the control of all
channel volumes and mutes from user space.

Signed-off-by: Jaska Uimonen <jaska.uimonen@intel.com>

### CODE REVIEW

**Comment by juimonen** [comment]
related driver PR: https://github.com/thesofproject/linux/pull/2469

**Comment by kv2019i** [comment]
@juimonen Does this impact the existing UCM files?

**Comment by juimonen** [comment]
@kv2019i yes good question. I don't know do we currently use the actual 4-channel topology anywhere? X1 could be one candidate. So the dmic pcm or control names are not changing, but the channel count is. 

We might be using 4 channel dmic topology in X1, but as our current topology uses 2 channel controls and asoc actually restricts the channel count to 2 channel, we will have 2 channel control in user space. So UCM works with this, but you are only controlling 2 dmic channels with this -> existing current bug (you can't totally mute the array for example).

So we should really see can we now have conditional channel count for dmic in ucm, and also what will pulseaudio do with 4 channel input volume.. (?)

Anyway we need to do something to the 4ch dmic volume.

**Comment by singalsu** [comment]
This PR causes with old kernel a regression for 4ch capture. I'm very sorry to report that since this is the right thing to do if ALSA/ASoC side would work.

There's many volume instances so it's hard to follow them in boot. This way the control to instance Dmic can be seen easily: 

`arecord -Dhw:0,1 -f S16_LE -c 4 -r 48000 -d 10 rec01.wav & sleep 5; amixer cset 'name=Dmic0 Capture Volume' 100%`

The result trace snippet is:

```
[579079297.864583] (  4425785.000000) c0 pga          7.42  ../audio/volume/volume.c:561  volume_ctrl_set_cmd(), channel = 0, value = 655434
[579079302.864583] (        5.000000) c0 pga          7.42  ../audio/volume/volume.c:561  volume_ctrl_set_cmd(), channel = 1, value = 655434
[579079306.093750] (        3.229167) c0 pga          7.42  ../audio/volume/volume.c:561  volume_ctrl_set_cmd(), channel = 2, value = 0
[579079309.270833] (        3.177083) c0 pga          7.42  ../audio/volume/volume.c:561  volume_ctrl_set_cmd(), channel = 3, value = 0

```
So this would cause a regression with muted audio for any channels above stereo, if using current kernel without Jaska's patch there.

**Comment by juimonen** [comment]
@singalsu yep this is the case.

1) So with this new topology and old kernel asoc interprets this as mono control, and god knows what channels are set
2) OTH with old kernel old topology we have an existing bug of not changing the last 2 channels volume
3) With new kernel and and new topology this would be solved, but we still need to check ucm and pulseaudio behavior

So all and all I don't see too much choices than going with 3, or then hardwire the dmic multi-channel volume somehow to firmware (so that mono/stereo control would always change all channels). The hardwire options seems also technically quite hackish...

**Comment by juimonen** [comment]
just a note to reviewers, this PR is only affecting topologies with 4 channel dmic configuration. 2 channel dmic topologies and all the rest of the topologies should remain the same. At least that is the intention...

**Comment by lgirdwood** [comment]
@juimonen @kv2019i does this need a kernel "fix" patch that can be back ported to older stable kernels ? or is the kernel change too complex here ?

**Comment by juimonen** [comment]
@lgirdwood yes this needs a kernel patch, it is in PR https://github.com/thesofproject/linux/pull/2469. I think it can be backported? The change is currently in sof only, although there's a discussion, what is the best way to handle this

But yes, the situation is following:
- We currently have a bug in 4ch dmic topologies -> you can only change the volume of 2 channels (so for example mute is problematic because 2 dmic might remain open)
- If this topology PR is currently used with old kernels (without the mentioned driver PR) the volume control will be mono, because of asoc mono/stereo restriction, so if channel count > 2, asoc will decide it is mono... so we will have the above bug in a little bit different way

I really don't know how this would be best handled and synced with upstream...

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
> @lgirdwood yes this needs a kernel patch, it is in PR [thesofproject/linux#2469](https://github.com/thesofproject/linux/pull/2469). I think it can be backported? The change is currently in sof only, although there's a discussion, what is the best way to handle this

ok, I think this pends on the kernel discussion. Can you ping here when kernel part is resolved and we can then merge.

**Comment by juimonen** [comment]
@lgirdwood kernel part is merged now. For the use case to work through whole sw stack we need this PR and upstream change to pulseaudio:
https://gitlab.freedesktop.org/pulseaudio/pulseaudio/-/merge_requests/382
